### ISSUE TITLE
usage of threads

### ISSUE THREAD

**Issue by levBagryansky** [objectionary/eo-threads#4]
@mximp 
Please rate my suggestion. I came with 2 usage of threads. 

1. Its enough ordinary usage. You just run the thread and can use several attributes to get its state: join, is-alive, etc .
```
[] > usage1
  QQ.threads.thread > my-thread
    slow-obj
  seq > @
    run.
      my-thread
    do-smth
    join.
      my-thread
```
2. The fist usage may have too imperative style. So its alternative. You can declarative some threads in one place and run it where you need. It will return true after finish.
```
 [] > usage2
  QQ.threads.thread-set > my-thread-set
    slow-obj1
    slow-obj2
    slow-obj3
  run. > @
    my-thread-set
```

**Comment by levBagryansky** [comment]
Also we can create something like `lock` or `mutex` for the simplest thread coordination

**Comment by levBagryansky** [comment]
@yegor256 , please criticize

**Comment by levBagryansky** [comment]
@yegor256 , my first variant is almost that, but it does not use `[x]`
```
      QQ.threads.thread > t
        [x]
          very-slow-object > @
```
I think this usage is strange. Why do we need [x]? We can use slow-obj as an argument

**Comment by yegor256** [comment]
@levBagryansky here `x` is the thread. Passing it into the object will give more control to it.

**Comment by levBagryansky** [comment]
@yegor256 , is thread an argument for thread? Can you give an example how where x is using?

**Comment by mximp** [comment]
@levBagryansky what does 'run a thread' means in EO? We have only dataization in place.

**Comment by levBagryansky** [comment]
@mximp , run returns `TRUE` and dataizes the thread in parallel

**Comment by yegor256** [comment]
@levBagryansky I think `start` is a better name, since `run` implies that something will run and we will wait. `start` means that we only trigger the start but don't wait.

**Comment by levBagryansky** [comment]
@mximp , it dataizes the argument

**Comment by yegor256** [comment]
@levBagryansky in Java a thread can get access to itself using `Thread.currentThread()`. something similar would be nice to have in EO

**Comment by levBagryansky** [comment]
> @levBagryansky I think `start` is a better name, since `run` implies that something will run and we will wait. `start` means that we only trigger the start but don't wait.

ok

**Comment by yegor256** [comment]
@levBagryansky probably not, since there is no thread but only a process

**Comment by levBagryansky** [comment]
@yegor256 , if we describe a slow-obj as an argument, we can to use attributes of threads

**Comment by mximp** [comment]
@yegor256 @levBagryansky 

We also may want to supply some approach to work with dataized underlying object:
```
[] > running-code-in-a-separate-thread
  QQ.threads.thread > t
    []
      slow-obj > @
    [obj]
      QQ.io.stdout > @
        QQ.txt.sprintf
          "Result: %d"
          obj

  seq > @
    t.start
    some-other-code
```

**Comment by levBagryansky** [comment]
@mximp , if I understood correctly, the tread will firstly dataize 
```
    []
      slow-obj > @
```
and then 
```
    [obj]
      QQ.io.stdout > @
        QQ.txt.sprintf
          "Result: %d"
          obj
```
Why can we not just dataize it in `seq`?

**Comment by yegor256** [comment]
@mximp maybe this way is better:

```
[] > running-code-in-a-separate-thread
  QQ.threads.thread > t
    []
      slow-obj > @
  seq > @
    t.start
    t.after
      [obj]
        QQ.io.stdout > @
          QQ.txt.sprintf
            "Result: %d"
            obj
    some-other-code
```

The method `.after` will tell the thread what to do when it's finished.

**Comment by levBagryansky** [comment]
> @levBagryansky here `x` is the thread. Passing it into the object will give more control to it.
> @levBagryansky in Java a thread can get access to itself using `Thread.currentThread()`. something similar would be nice to have in EO
@yegor256 so `x` is a Thread.currentThread()?

**Comment by yegor256** [comment]
@levBagryansky in Java implementation of `QQ.threads.thread`, yes.

**Comment by mximp** [comment]
> @mximp maybe this way is better:
> 
> ```
> [] > running-code-in-a-separate-thread
>   QQ.threads.thread > t
>     []
>       slow-obj > @
>   seq > @
>     t.start
>     t.after
>       [obj]
>         QQ.io.stdout > @
>           QQ.txt.sprintf
>             "Result: %d"
>             obj
>     some-other-code
> ```
> 
> The method `.after` will tell the thread what to do when it's finished.

@yegor256 In this case what if the thread finishes earlier than `t.after` is dataized? Having start of a thread and post action definition as separate steps we have no guaranty post-action is exectuted.

**Comment by levBagryansky** [comment]
@yegor256 How can we use result from this thread?
```
QQ.threads.thread > a
  []
    1.plus > @
      1
```
Is it illegal to write 
```
[] > multiple
  QQ.threads.thread > a
    [x]
      2.plus 2 > @
  QQ.threads.thread > b
    [x]
      1.plus 1 > @
  a.times > @
    b

```
?

**Comment by levBagryansky** [comment]
@yegor256 so must it work?
```
[] > multiple
  QQ.threads.thread > a
    [x]
      2.plus 2 > @
  QQ.threads.thread > b
    [x]
      1.plus 1 > @
  a.start
  b.start
  a.times > @
    b
```

**Comment by levBagryansky** [comment]
@yegor256 
In the sample above when `b` will be dataized `b.start` can be ignored. It will dataize the line where `b` was declarated

**Comment by mximp** [comment]
> @yegor256 How can we use result from this thread?
> 
> ```
> QQ.threads.thread > a
>   []
>     1.plus > @
>       1
> ```
> 
> Is it illegal to write
> 
> ```
> [] > multiple
>   QQ.threads.thread > a
>     [x]
>       2.plus 2 > @
>   QQ.threads.thread > b
>     [x]
>       1.plus 1 > @
>   a.times > @
>     b
> ```
> 
> ?
@levBagryansky No, it won't work like this. In this form you cannot use the result of a thread. It will be just dataized in a separate thread.

**Comment by mximp** [comment]
> @yegor256 so must it work?
> 
> ```
> [] > multiple
>   QQ.threads.thread > a
>     [x]
>       2.plus 2 > @
>   QQ.threads.thread > b
>     [x]
>       1.plus 1 > @
>   a.start
>   b.start
>   a.times > @
>     b
> ```

@levBagryansky I believe this code is illegal in EO. Nothing will trigger `*.start` dataization here.

**Comment by levBagryansky** [comment]
@yegor256
I have a suspicion this may not work: when it will try to dataize `a.start `it will firstly compute `a`. And since `a` was not started yet it will return error.
Also I noticed  the set of attributes has `a` here depends on was it alreade started(then there are `slow-object` atrbutes) or not(the there are attributes of `thread`). May be it is better better to intoduce `result` that is dataized `slow-obj` to avoid these uncertainty?

**Comment by yegor256** [comment]
@levBagryansky `a.start` will not compute `a.@`

**Comment by levBagryansky** [comment]
@yegor256 Is it ok that both `a.is-running` and `a,times` are legal here?
I understand correctly, when it see `thread.times` does not exist it tries to call `a.@.times`, and that is why executes `a.@`?

**Comment by yegor256** [comment]
@levBagryansky yes, you understand it correctly. yes, both `a.is-running` and `a.times` are legal.

**Comment by levBagryansky** [comment]
@yegor256 thanks!

**Comment by yegor256** [comment]
@levBagryansky actually, on the second thought, maybe it's better to change the code to this:

```
[] > multiple
  QQ.threads.thread > a
    2.plus 2
  QQ.threads.thread > b
    1.plus 1
  seq > @
    a.start
    b.start
    a.result.times b
```

This `a.result` is safer than implicit dataization of `a`, which may lead to collisions of names (`start` may exist in the result of dataization)

**Comment by levBagryansky** [comment]
@yegor256 I like this variant.
So `a.join` waits for argument's dataization and returns it?

**Comment by mximp** [comment]
> @levBagryansky actually, on the second thought, maybe it's better to change the code to this:
> 
> ```
> [] > multiple
>   QQ.threads.thread > a
>     2.plus 2
>   QQ.threads.thread > b
>     1.plus 1
>   seq > @
>     a.start
>     b.start
>     a.join.times b
> ```
> 
> This `a.join` is safer than implicit dataization of `a`, which may lead to collisions of names (`start` may exist in the result of dataization)

`a.join.times b` might not work as `b` will be a thread here. 
It should be `a.join.times b.join` probably.
